### DISCUSSION TITLE
creating a conditional authentication dependency

### DISCUSSION THREAD

**Discussion by havok2063** [fastapi/fastapi#8445] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends, Request
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="login")

async def release(release: str = 'A') -> str:
    """ Dependency to specify a release query parameter """
    return release

async def check_auth(request: Request, release = Depends(release)):
    release = release or 'A'
    if release != 'A':
        return None
    return await oauth2_scheme(request)

async def set_auth(token: str = Depends(check_auth), release = Depends(release)): 
    return {"token": token, 'release': release}

app = FastAPI(title='API', description='The API', dependencies=[Depends(set_auth)])

@app.get("/", summary='Hello World route')
def hello(release = Depends(release)):
    return {"Hello": "Hello World", 'release': release}

@app.post("/login")
async def login(form_data: OAuth2PasswordRequestForm = Depends()):
    # standard normal login stuff
    # ...
    return {"access_token": 'blah', "token_type": "bearer"}
```

### Description

I'm trying to create an API with a conditional authentication system.  This API will serve data bundled into releases, e.g. A, B, C.   Releases B, C are public and A is proprietary.  All the API routes should work for all the data, but data from release A needs to be behind authentication.  Routes should work without an auth prompt for all releases not from A.  I've created a standard OAuth2PasswordBearer system and wrapped it inside a `check_auth` dependency.  The example code provided works.  I can do

```
requests.get('http://localhost:8000/?release=B').json()
{"Hello": "Hello World", 'release': 'B'}

requests.get('http://localhost:8000/?release=A').json()
{'detail': 'Not authenticated'}
```
but the use of `await oauth2_scheme(request)` doesn't feel very FastAPI-ish to me.  My question is does this solution make sense, or is there a better way, more in-align with best practices for FastAPI?  Also, it looks like the Authorize option in my rendered documentation (`localhost:8000/docs`) is gone.  Is there a way to get that to appear with this implementation?

### Operating System

macOS

### FastAPI Version

0.74.1

### Python Version

3.8

**Comment by havok2063** [comment]
I've simplified the code a bit by subclassing `OAuth2PasswordBearer` directly.  This solves the issue with the Authorize option not appearing in the docs, while still giving me the conditional auth on release parameter.

```python
class MyAuthPasswordBearer(OAuth2PasswordBearer):
    
    async def __call__(self, request: Request, release: str = Depends(release)):
        self.release = release or "A"
        if self.release != 'A':
            return None
        await super().__call__(request)
        
oauth2_scheme = MyAuthPasswordBearer(tokenUrl="login")

async def set_auth(token: str = Depends(oauth2_scheme), release = Depends(release)): 
    return {"token": token, 'release': release}

app = FastAPI(title='API', description='The API', dependencies=[Depends(set_auth)])

@app.get("/", summary='Hello World route')
def hello(release = Depends(release)):
    return {"Hello": "Hello World", 'release': release}
```

**Comment by adriangb** [comment, marked as answer]
I think there's two options you could explore

## Make this two separate endpoints

Then you can have different auth dependencies in each

## Set `auto_error=False` and manually error


```python
auth = OAuth2PasswordBearer(..., auto_error=False)


async def check_auth(release: str = Depends(release), token: Optional[str] = Depends(auth)) -> None:
    if release == "A" and not token:
        raise HTTPException(401)
```
